lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+ */
+class Gauge {
+  /**
+   * The metric to report to.
+   * @property {Metric} metric
+   */
+
+  /**
+   * The value of the gauge.
+   * @property {number} value
+   */
+
+  /**
+   * @param {Metric} metric The metric to report to.
+   */
+  constructor(metric) {
+    if (!(metric instanceof Metric)) {
+      throw new ERR_INVALID_ARG_TYPE('metric', ['Metric'], metric);
+    }
+    this.metric = metric;
+    this.value = 0;
+  }
+
+  /**
+   * Set the gauge value.
+   * @param {number} value The value to set the gauge to.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  reset(value = 0, meta) {
+    this.value = value;
+    this.metric.report(value, meta);
+  }
+
+  /**
+   * Apply a delta to the gauge.
+   * @param {number} value The delta to apply to the gauge.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  applyDelta(value, meta) {
+    this.reset(this.value + value, meta);
+  }
+}
+
+/**
+ * Number of events per interval. This will report at every mark() call,
+ * but will report an aggregate value if a sliding window of marks which
+ * occurred within the interval period.
+ */
+class Meter extends Gauge {
+  #window;
+
+  /**
+   * The metric to report to.
+   * @property {Metric} metric
+   */
+
+  /**
+   * The interval in milliseconds to aggregate marks over.
+   * @property {number} interval
+   */
+
+  /**
+   * Construct a new meter.
+   * @param {Metric} metric The metric to report to.
+   * @param {number} interval The interval in milliseconds to aggregate marks over.
+   */
+  constructor(metric, interval) {
+    super(metric);
+    if (typeof interval !== 'number' || interval <= 0) {
+      throw new ERR_INVALID_ARG_TYPE('interval', ['number'], interval);
+    }
+    this.interval = interval;
+    this.#window = [];
+  }
+
+  /**
+   * Mark an event in the meter.
+   * @param {number} [n] The number of events to mark. Defaults to 1.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  mark(n = 1, meta) {
+    if (!this.metric.shouldReport) return;
+
+    if (typeof n === 'object') {
+      meta = n;
+      n = 1;
+    }
+
+    const now = performance.now();
+    this.#window.push({ value: n, time: now });
+
+    let { value } = this;
+    while (this.#window.length && (now - this.#window[0].time) > this.interval) {
+      const cached = this.#window.shift();
+      value -= cached.value;
+    }
+
+    value += n;
+    this.reset(value, meta);
+  }
+}
+
+/**
+ * An increasing or decreasing value.
+ */
+class Counter extends Gauge {
+  /**
+   * Increment the counter. Negative values invert to positive.
+   * @param {number} [n] The amount to increment the counter by. Defaults to 1.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  increment(n = 1, meta) {
+    if (!this.metric.shouldReport) return;
+
+    if (typeof n === 'object') {
```

Comment:
should these allow for `null`?

Reply:
Not sure there's a reason one would ever provide explicit null there. 🤔